rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -3037,27 +3037,28 @@ static rmw_ret_t do_for_node_user_data(
   return do_for_node(node_impl, f);
 }
 
-extern "C" rmw_ret_t rmw_get_node_names(
+extern "C" rmw_ret_t rmw_get_node_names_impl(
+  std::regex re,
   const rmw_node_t * node,
   rcutils_string_array_t * node_names,
-  rcutils_string_array_t * node_namespaces)
+  rcutils_string_array_t * node_namespaces,
+  rcutils_string_array_t * security_contexts)
 {
   RET_WRONG_IMPLID(node);
   auto node_impl = static_cast<CddsNode *>(node->data);
-  if (rmw_check_zero_rmw_string_array(node_names) != RMW_RET_OK ||
-    rmw_check_zero_rmw_string_array(node_namespaces) != RMW_RET_OK)
+  if (RMW_RET_OK != rmw_check_zero_rmw_string_array(node_names) ||
+    RMW_RET_OK != rmw_check_zero_rmw_string_array(node_namespaces))
   {
     return RMW_RET_ERROR;
   }
 
-  std::vector<std::pair<std::string, std::string>> ns;
-  const auto re = std::regex("^name=([^;]*);namespace=([^;]*);", std::regex::extended);
+  std::vector<std::tuple<std::string, std::string, std::string>> ns;
   auto oper =
     [&ns, re](const dds_builtintopic_participant_t & sample, const char * ud) -> bool {
       std::cmatch cm;
       static_cast<void>(sample);
       if (std::regex_search(ud, cm, re)) {
-        ns.push_back(std::make_pair(std::string(cm[1]), std::string(cm[2])));
+        ns.push_back(std::make_tuple(std::string(cm[1]), std::string(cm[2]), std::string(cm[3])));
```

Comment:
@ivanpauno this bit is still assuming that there's a third group, regardless of `security_contexts` value.

Is having a `security_context` stanza in `ud` mandatory? If not, why not simply making it optional in the regex? Then, `rmw_get_node_names(...)` reduces to `rmw_get_node_names_with_security_contexts(..., nullptr)`.

Reply:
This is wrong, thanks for detecting it!

Reply:
See https://github.com/ros2/rmw_cyclonedds/pull/106/commits/e09008b242ce747ee08ac0075ed8823d1253fa57.
Based on the [docs](https://en.cppreference.com/w/cpp/regex/match_results/operator_at), the previous code wasn't wrong. I actually ran tests, and it was working (forget about the detail that used the same regex both time and it wasn't going to match with old implementations).

If you have a better way to write the regex, please provide it.

Reply:
Ok, finally fixed in https://github.com/ros2/rmw_cyclonedds/pull/106/commits/d213440d429a2638de95ead7f2dee1bdc03f229c.
I will rely on the `Eloquent` and `Dashing` checks here for validity, because the same code is exercised there.